File: examples/remote-colony/agent.ts
/**
 * agent.ts — a Stigmergy agent living in its own process.
 *
 *   STIGMERGY_URL=http://127.0.0.1:8787 \
 *   STIGMERGY_TOKEN=stg_… \
 *   npx tsx examples/remote-colony/agent.ts
 *
 * Notice what this file does NOT import: the colony definition, the database,
 * any signal schemas. It connects with a token, asks the server what roles it
 * has, and acts within them. The same binary works as a Reporter or a Triager
 * depending on which token you hand it — the behavior is selected from the
 * session, not hard-wired.
 *
 * In your own project this import is `from "stigmergy"`.
 */

import { type RemoteAgentSession, connect } from "../../src/index.js";

const SEED_BUGS = [
  { title: "login-loops-on-SSO",        component: "auth",     severity: 1, body: "infinite redirect after IdP callback" },
  { title: "slow-dashboard-on-firefox", component: "frontend", severity: 2, body: "TTI > 8s on cold load, firefox only" },
  { title: "duplicate-welcome-email",   component: "backend",  severity: 2, body: "new signups receive two emails" },
  { title: "typo-in-footer",            component: "frontend", severity: 3, body: "says 'Copywrite' instead of 'Copyright'" },
  { title: "payment-webhook-500s",      component: "backend",  severity: 1, body: "Stripe retries for hours, no ingress" },
];

async function main(): Promise<void> {
  const url = process.env.STIGMERGY_URL ?? "http://127.0.0.1:8787";
  const token = process.env.STIGMERGY_TOKEN;
  if (!token) throw new Error("Set STIGMERGY_TOKEN (printed by server.ts on startup).");

  const session = await connect({ url, token });
  const roles = session.roles.map((r) => r.name);
  console.log(`connected as ${session.agentId} — roles: ${roles.join(", ")}`);
  if (session.charter) console.log(`charter: ${session.charter.trim()}`);

  process.on("SIGINT", () => {
    session.close();
    process.exit(0);
  });

  if (roles.includes("Reporter")) {
    await session.run(reporterTick, { intervalMs: 1500 });
  } else if (roles.includes("Triager")) {
    await session.run(triagerTick, { intervalMs: 1000 });
  } else {
    console.log("no known role to enact; idling.");
  }
}

async function reporterTick(s: RemoteAgentSession): Promise<void> {
  const filed = await s.as("Reporter").view();
  const seen = new Set(filed.map((b) => b.payload.title as string));
  const next = SEED_BUGS.find((b) => !seen.has(b.title));
  if (!next) return;
  console.log(`[reporter] file "${next.title}" (sev ${next.severity})`);
  await s.as("Reporter").deposit("reported_bug", { ...next, claimed_by: null, claimed_until: null });
}

async function triagerTick(s: RemoteAgentSession): Promise<void> {
  const queue = await s.as("Triager").view();
  const target = queue[0];
  if (!target) return;
  if (!(await s.as("Triager").tryClaim(target.id, { until: "2m" }))) return;

  const title = target.payload.title as string;
  // Postgres numeric columns arrive as strings on the wire; coerce explicitly.
  const severity = Number(target.payload.severity);
  const verdict = title.includes("typo") ? "invalid" : severity === 3 ? "duplicate" : "confirm";
  const boost = verdict === "confirm" ? 4 - severity : 0;

  console.log(`[${s.agentId}] ${verdict.padEnd(9)} "${title}" (strength ${(target.strength ?? 0).toFixed(2)})`);
  await s.as("Triager").deposit("triage_note", {
    bug_id: target.id,
    bug_title: title,
    verdict,
    body: `auto-triage by ${s.agentId}`,
    recommended_boost: boost,
  });
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/remote-colony/agent.py
#!/usr/bin/env python3
"""agent.py — a Stigmergy agent in ~70 lines of Python standard library.

    STIGMERGY_URL=http://127.0.0.1:8787 \\
    STIGMERGY_TOKEN=stg_... \\
    python3 examples/remote-colony/agent.py

There is no Stigmergy Python package. There doesn't need to be: the colony is
reachable over plain JSON-HTTP, so an agent in any language is a few requests.
This one enacts the Triager role. Point it at a triager token and it competes
for claims against the TypeScript triager exactly the same way — the medium
doesn't know or care what language deposited a signal.
"""

import json
import os
import time
import urllib.request
import urllib.error

URL = os.environ.get("STIGMERGY_URL", "http://127.0.0.1:8787").rstrip("/")
TOKEN = os.environ.get("STIGMERGY_TOKEN")
if not TOKEN:
    raise SystemExit("Set STIGMERGY_TOKEN (printed by server.ts on startup).")


def call(route, body):
    """POST {body} to /v1{route}; return the parsed JSON response."""
    req = urllib.request.Request(
        f"{URL}/v1{route}",
        data=json.dumps(body).encode("utf-8"),
        headers={
            "content-type": "application/json",
            "authorization": f"Bearer {TOKEN}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req) as res:
            return json.loads(res.read().decode("utf-8"))
    except urllib.error.HTTPError as err:
        detail = json.loads(err.read().decode("utf-8")).get("error", {})
        raise SystemExit(f"{err.code} {detail.get('code')}: {detail.get('message')}")


def main():
    session = call("/session", {})
    roles = [r["name"] for r in session["roles"]]
    print(f"connected as {session['agentId']} — roles: {', '.join(roles)}")
    if "Triager" not in roles:
        raise SystemExit("This Python agent enacts the Triager role; hand it a triager token.")

    while True:
        queue = call("/view", {"role": "Triager"})["signals"]
        if not queue:
            time.sleep(1.0)
            continue

        target = queue[0]
        claimed = call("/claim", {"role": "Triager", "signalId": target["id"], "until": "2m"})
        if not claimed["claimed"]:
            continue  # another triager won the race

        payload = target["payload"]
        # Postgres numeric columns arrive as strings on the wire (same as the
        # in-process API, where JS would coerce them silently). Coerce here.
        title, severity = payload["title"], int(payload["severity"])
        verdict = "invalid" if "typo" in title else "duplicate" if severity == 3 else "confirm"
        boost = 4 - severity if verdict == "confirm" else 0

        strength = target.get("strength", 0.0)
        print(f"[{session['agentId']}] {verdict:<9} \"{title}\" (strength {strength:.2f})")
        call("/deposit", {
            "role": "Triager",
            "type": "triage_note",
            "payload": {
                "bug_id": target["id"],
                "bug_title": title,
                "verdict": verdict,
                "body": f"auto-triage by {session['agentId']} (python)",
                "recommended_boost": boost,
            },
        })
        time.sleep(1.0)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass
